### DISCUSSION TITLE
Github Container registry (ghcr.io) packages not appearing in api

### DISCUSSION THREAD

**Discussion by christian-korneck** [community/community#26627] [Packages]
<p>Is there any way to query the ghcr type packages using the github api (rest or graphql)?<br>
The ghcr docs have an <a href="https://docs.github.com/en/packages/managing-container-images-with-github-container-registry/deleting-a-container-image#deleting-a-version-of-a-package-with-graphql" rel="nofollow noopener">example of how to delete a ghcr container image version with graphql</a>. That mutation needs the version id of the package - <strong>but how can I get that ID (or other infos about the package)?</strong></p>
<p>This graphql query seems to only lists the “old” type of Docker packages (<a href="http://docker.pkg.github.io" rel="nofollow noopener">docker.pkg.github.io</a>):</p>
<pre><code class="lang-auto">{
  viewer {
    login
    packages(first: 100) {
      edges {
        node {
          id
          name
          packageType
        }
      }
    }
  }
}
</code></pre>
<p>Am I looking in the wrong place? Thanks a lot for any advice!</p>

**Comment by nohaapav** [comment]
<p>Having the same issue. Only old packages using <code>docker.pkg.github.com</code> domain are listed. Although these packages are listed in Github UI so there must be some query to fetch such information. Can you update api docs or share the correct query please ?</p>

**Comment by christian-korneck** [comment]
<p>anyone any idea? (even if it’s not possible yet to manage ghcr containers via the API it would be great to get a quick confirmation). Thanks!!</p>

**Comment by jcansdale** [comment]
<p>Hi <a class="mention" href="/christian-korneck">@christian-korneck</a>,</p>
<p>Sorry about the delay!</p>
<aside class="quote no-group" data-username="christian-korneck" data-post="3" data-topic="130133" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 christian-korneck:</div>
<blockquote>
<p>anyone any idea? (even if it’s not possible yet to manage ghcr containers via the API it would be great to get a quick confirmation). Thanks!!</p>
</blockquote>
</aside>
<p>I’m afraid it isn’t possible to manage ghcr containers via the GraphQL API. The previous registry version was heavily based on GraphQL. This is a very different beast and will likely have a different API.</p>

**Comment by nohaapav** [comment]
<p>Hi <a class="mention" href="/jcansdale">@jcansdale</a>,</p>
<p>is there any ETA for delivery ?</p>

**Comment by jcansdale** [comment, marked as answer]
<aside class="quote no-group" data-username="nohaapav" data-post="5" data-topic="130133">
<div class="title">
<div class="quote-controls"></div>
 nohaapav:</div>
<blockquote>
<p>is there any ETA for delivery ?</p>
</blockquote>
</aside>
<p>No specific ETA. I’ve confirmed that the team are working on a REST API for this.</p>
<p>Keep an eye on the change log:<br>
<a href="https://github.blog/changelog/" class="onebox" target="_blank" rel="nofollow noopener">https://github.blog/changelog/</a></p>
<p>I’ll try to report back here as well!</p>

**Comment by christian-korneck** [comment]
<p>thanks <a class="mention" href="/jcansdale">@jcansdale</a>, I think that answers my question.</p>

**Comment by den-is** [comment]
<p>same here :frowning:<br>
that makes migration from <strong><a href="http://docker.pkg.github.com" rel="noopener nofollow ugc">docker.pkg.github.com</a></strong> to <strong><a href="http://ghcr.io" rel="noopener nofollow ugc">ghcr.io</a></strong> painful for us.</p>

**Comment by WyriHaximus** [comment]
<p>Is there a timeline for when we might see this? With the DockerApocalypse coming on Monday this is quickly becoming more and more relevant.</p>

**Comment by benc-uk** [comment]
<p>Just hit this today when trying to create some automation to clean up old images from GHCR<br>
API support is essential</p>

**Comment by whitneyimura** [comment]
<p>:wave: <a class="mention" href="/benc-uk">@benc-uk</a> We’re actively talking about this exact use case, and a delete REST endpoint will likely be the first API we expose to our customers for Container Registry.  We are hoping to expose such an endpoint before end of year.</p>

**Comment by airtower-luna** [comment]
<p>Will you support the <a href="https://docs.docker.com/registry/spec/api/#deleting-an-image">Docker Registry API for image deletion</a>, or are you planning something GHCR specific? I think a generic solution would be preferable. Not that I’d mind an additional GHCR API for additional features, of course (especially if it offers something similar to <code>docker image prune</code>)! :wink:</p>

**Comment by derekbelrose** [comment]
<p>Hi Whitney,</p>
<p>Is there any update on this REST endpoint? I would, also, like to be able to modify permissions on an image and, if possible, trigger a GH action when an image has been pushed.</p>

**Comment by rubengees** [comment]
<p>Hello!</p>
<p>Is there any progress on this? The old repository-bound container package registry is deprecated and the new one does not show up in the GraphQL APIs yet. It would be really helpful to have this so we can automate package deletion. This is actively blocking us from migrating to the new registry and that in terms is blocking us from upgrading to K8s 1.20 which does only work with the new registry.</p>

**Comment by jcansdale** [comment]
<p>Hi <a class="mention" href="/rubengees">@rubengees</a>,</p>
<aside class="quote no-group" data-username="rubengees" data-post="14" data-topic="130133">
<div class="title">
<div class="quote-controls"></div>
 rubengees:</div>
<blockquote>
<p>Is there any progress on this? The old repository-bound container package registry is deprecated and the new one does not show up in the GraphQL APIs yet. It would be really helpful to have this so we can automate package deletion.</p>
</blockquote>
</aside>
<p>Yes, the new REST API has been released!</p>
<p>You can find it here:</p><aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/rest/reference/packages" target="_blank" rel="noopener nofollow ugc">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/rest/reference/packages" target="_blank" rel="noopener nofollow ugc">Packages - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/rest/reference/packages" target="_blank" rel="noopener nofollow ugc">//docs.github.com/en/rest/reference/packages</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Here is a blog post about package deletion (including public packages) for all registries:</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.blog/wp-content/uploads/2019/01/cropped-github-favicon-512.png?fit=32%2C32" class="site-icon" width="32" height="32">
      <a href="https://github.blog/changelog/2021-02-17-packages-delete-and-restore-for-all-packages/" target="_blank" rel="noopener nofollow ugc">The GitHub Blog</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://github.blog/wp-content/uploads/2019/03/blog-card.png?fit=5001%2C2626" class="thumbnail" width="690" height="362"></div>

<h3><a href="https://github.blog/changelog/2021-02-17-packages-delete-and-restore-for-all-packages/" target="_blank" rel="noopener nofollow ugc">Packages: delete and restore for all packages - GitHub Changelog</a></h3>

<p>Packages: delete and restore for all packages</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Please let me know if you have any further questions.</p>

**Comment by benc-uk** [comment]
<p>The docs state "GraphQL is not supported for containers or Docker images at <code>ghcr.io</code>"</p>
<p>So how are we supposed to bulk delete many old images in GH Container Registry ?<br>
Click for hours in web UI?</p>

**Reply by polarathene** [reply]
Not sure which docs page you were quoting at the time, link would have been helpful as GraphQL docs on packages doesn't seem to clearly indicate how unusable it is vs the REST API.

Roughly a week earlier than your comment here, [another comment in the discussion](https://github.com/orgs/community/discussions/26627#discussioncomment-3252602) shared the [news of REST API support for deleting packages](https://github.blog/changelog/2021-02-17-packages-delete-and-restore-for-all-packages/) (Feb 2021), yet the docs in 2026 [still have a disclaimer that support is experimental](https://docs.github.com/en/packages/learn-github-packages/deleting-and-restoring-a-package#packages-api-support).

**Comment by kc1r74p** [comment]
<p>You can use the REST Api like <a class="mention" href="/jcansdale">@jcansdale</a> mentioned, this way you can build any automation you want.</p>
<p>I also adjusted the packages action which can now remove old versions from ghcr.<br>
Related issue with PoC: <a href="https://github.com/actions/delete-package-versions/issues/28#issuecomment-804394142" class="inline-onebox" rel="noopener nofollow ugc">Cannot read property 'packages' of null · Issue #28 · actions/delete-package-versions · GitHub</a></p>
<p>However I just opened a thread here where I am wondering if we can also use the GITHUB_TOKEN instead of a Personal Access Token.</p><aside class="quote quote-modified" data-post="1" data-topic="170395">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/github-token-has-no-access-to-new-container-rest-apis/170395">Github token has no access to new container REST APIs</a> <a class="badge-wrapper  bullet" href="/c/code-to-cloud/github-packages/43"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #005cc5;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Any questions related to GitHub Packages and how to manage your packages; upload, download, and delete. Happy to see questions about our help docs and the core set of clients and services we support but also questions about configuring and using alternate clients are welcome.">GitHub Packages</span></a>
  </div>
  <blockquote>
    Hello, 
I just saw the changelog entry which now introduced access for the container registry. 
There it mentions that the secrets.GITHUB_TOKEN now has full access including reading and deleting for the ghcr registry. 
However, switching from PAT to the GITHUB_TOKEN and adding permissions (push works fine with new token), now my custom action to list and then delete containers from <a href="http://ghcr.io" rel="noopener nofollow ugc">ghcr.io</a> throws: 

HttpError: Resource not accessible by integration 

when using following API with a non PAT token…
  </blockquote>
</aside>

<p>Best regards<br>
Patrick</p>
